sklearn/cluster/_bisect_k_means.py

```diff
@@ -0,0 +1,708 @@
+                labels[labels_left[indexes_left]] = n_iter + 1
+
+                if self.verbose:
+                    print(f"Centroid Found: {centers[higher_sse_index]}")
+                break
+
+            # Pass for further split data, weights and labels left
+            data_left = data_left[indexes_left]
+            weights_left = weights_left[indexes_left]
+            labels_left = labels_left[indexes_left]
+
+        self.labels_ = labels
+        self.cluster_centers_ = np.asarray(centroids)
+
+    def _bisect_by_biggest_sse(self, X, init, random_state, sample_weight):
+        """ Performs Bisecting K-Means, which splits always cluster with
+         biggest SSE from all calculated
+
+            .. note:: All of passed parameters must be pre-calculated
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            Training instances to cluster.
+
+        init : {'k-means++', 'random'}, callable or ndarray of shape \
+                (n_clusters, n_features)
+                Method for initialization.
+
+        random_state : int, RandomState instance or None, default=None
+            Determines random number generation for centroid initialization.
+
+        sample_weight : array-like of shape (n_samples,), default=None
+            The weights for each observation in X.
+
+
+        Returns
+        -------
+        centers : ndarray of shape (n_clusters, n_features)
+            The cluster centers.
+        """
+
+        label_indexes = np.arange(X.shape[0])
+
+        centers_dict = {
+            0: {'data': X, 'weights': sample_weight, 'sse': None,
+                'centroid': None, 'label_indexes': label_indexes}
+        }
+
+        last_center_id = 0
+
+        for n_iter in range(self.n_clusters - 1):
+
+            biggest, _ = max(centers_dict.items(), key=lambda x: x[1]['sse'])
+
+            # Perform Bisection
+            centers, labels = self._bisect(centers_dict[biggest]['data'],
+                                           init,
+                                           centers_dict[biggest]['weights'],
+                                           random_state)
+
+            # Check SSE (Sum of Squared Errors) of each of computed centroids.
+            # SSE is calculated with distances between data points
+            # and assigned centroids
+            errors = self._compute_bisect_errors(
+                centers_dict[biggest]['data'],
+                centers, labels,
+                centers_dict[biggest]['weights'])
+
+            lower_sse_index = 0 if errors[0] < errors[1] else 1
+            higher_sse_index = 1 if lower_sse_index == 0 else 0
+
+            higher_labels = (labels == higher_sse_index)
+            lower_labels = (labels == lower_sse_index)
+
+            # Add both centroids to dict
+            centers_dict[last_center_id + 1] = {
+                'data': centers_dict[biggest]['data'][lower_labels],
```

Comment:
Storing `data` seems unnecessary (assuming it makes a copy), since you already store the samples indices (`label_indexes` in your code).

Reply:
Yeah, when labels_left came it it becomes unnecessary, so storage of `data` and `weights` will be removed